accepted/2021/experimental-features/experimental-features.md

```diff
@@ -0,0 +1,380 @@
+support is sufficiently bootstrapped, but it's the most important part in the
+validation of the feature set. Practically, this means a good chunk of the
+feature will only be available halfway through a release cycle, which greatly
+reduces the time window for gathering feedback from other parts of the product
+as well as from customers using preview builds of the upcoming .NET version.
+
+Sadly, preview feedback isn't always sufficient to judge how well a feature
+works because
+
+1. Much fewer people use preview builds of .NET and
+2. Many people don't use .NET previews for their production projects, thus
+   greatly reducing the breadth & depth at which previews are being exercised.
+
+This document is exploring a model where RTM quality builds of .NET can contain
+features that are considered experimental.
+
+The goal is to ensure that customers can discover those features & have an easy
+way to play with them while simultaneously being fully aware that these features
+aren't support yet and thus might change between versions of .NET, just like
+they would in preview builds.
+
+## Scenarios and User Experience
+
+### Enabling an experimental runtime feature
+
+Maddy is building a Tye application. She reads in a blog post that the .NET
+runtime team is experimenting with a new GC mode, code named FancyGC.
+
+She wants to see how this improves the performance of her backend. She's glad to
+see that the new GC mode was shipped as part of .NET 6, so she doesn't need to
+figure out how to run a new version of .NET in her CI lab. So she creates a
+feature branch and enables the new GC mode simply by changing the service's
+project file as follows:
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+    <ExperimentalFeatures>FancyGC</ExperimentalFeatures>
+  </PropertyGroup>
+
+</Project>
+```
+
+She then submits a draft PR against `main` to trigger their CI/CD pipeline which
+also runs some performance tests.
+
+### Enabling an experimental language feature
+
+During lunch, Adriana discusses a design problem with one of the architects who
+mentions that the upcoming static interface members in C# might be able to solve
+this issue. Adriana does a quick web search and finds a blog post from the C#
+team that showcases how this feature might look like. She wonders how she can
+plan with that. Optimistic, she copy & pastes the sample code into her project
+which immediately generates a compilation error:
+
+> error: static interface members are an experimental feature of C# 9.0 and must
+> be turned on explicitly.
+
+The error displays a light-bulb in Visual Studio, suggesting to turn the feature
+on, which modifies her project file as follows:
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+    <ExperimentalFeatures>StaticInterfaceMembers</ExperimentalFeatures>
```

Comment:
The compiler only supports a preview mode for all features. We do not have the infrastructure for previewing individual features. As discussed we tried this in the past and it's an unmaintainable mess because the testing matrix gets out of control pretty much immediately.

Reply:
Having to maintain a test matrix combination for all of the different language features we preview would be a massive cost increase. It would severely limit the amount of previewing we can do compared to what we do today

Reply:
This is an interesting contrast to the runtime level work, where we generally have multiple experiments in place, some of which last several releases, and the typical characteristic of those experiments is that they are of ok quality, but may conflict in their implementation such that we cannot enable all experiments at once. I suspect it comes down to what is the expected quality bar for experiments, and what is their purpose. We solve the experimental test matrix problem by only testing with a single experiment enabled at once, and attempts to use them in combination is quite risky.

Reply:
Our customers don't differentiate on a feature by feature basis, they just expect to use C# Next. Once they opt in they want the features and the interaction between them. This is why we test them together because that is the expected customer use case. 

The expected quality bar is quite high here. The reliability of our implementation directly impacts the quality of Visual Studio. A huge hurdle we had to cross in even doing preview features in RTM builds was demonstrating that we could maintain quality no matter how the flag was toggled or what the user typed in the editor. 

Yes the expectation for preview features in VS is less than that of RTM features. If corner cases have extra diagnostics, bad formatting, or even if we emit bad code that's all expected. Pretty much everything else though (performance, watson, non-fatal watson, etc ...) is held to virtually the same bar.

Reply:
I agree with Jared and Jonathan. I also think the separate discussions here about language vs cross-cutting features are part of the issue.  The runtime has been (at least somewhat) successful in having opt-in experimental features because they in general don't require any code changes on the part of the developer, or even rebuilds: should tiered JIT be enabled, should stack allocation of objects be enabled, should PGO be enabled, should async value task methods cache, and so on... all of these things can be changed simply via environment variables.  But the kinds of features that involve public APIs and language features do demand code changes, do entail binaries distributed with the experiment baked in, at which point it's a whole different ball game.

It's perfectly reasonable for the runtime to continue shipping experiments the way it has, assuming they're isolated from anything else in the stack as they generally have been.  But for anything that's visible via language or library, I think we should have a single opt-in mode that enables using all preview features from the language or library (and whatever runtime support is necessary to make them work).  And we should control them in a single manner: I don't have a strong opinion about whether it's a command-line switch or a property in a project file or both, but we already have langpreview, and at present that seems good enough.  It's not language vs library vs runtime; it's one stack, with one preview mode.

Reply:
Would it make sense to enable another level of preview called "experimental" - the way I personally see the problem is that we, the generall mass, cannot really play with anything that is not included in preview. I mean all new fancy features are only here for us if they *actually get confirmed for the C# vnext* and thus added to preview. Experimental langversion could help with shipping features, that, for example may not require runtime support like ParameterNullChecking or fancy Attributes, Discriminated Unions and so on. Because honestly - coding those features is in no way as hard as maintaining consistency of a language and not letting it go awry. In no way "experimental" would mean that these features will get into preview and maybe C# team will have a hard time deciding that "feature X" is not for C# after all even when some people like it. I believe many features could be added to such "experimental" version of a language and people could bring real, valuable partical experience on top of existing long discussions. "Experimental" features could then be updated with new releases of .Net SDK. It seems logical because it's a bleeding edge stuff anyway.

Reply:
> The runtime has been (at least somewhat) successful in having opt-in experimental features

I think these were mostly enhancement and not full-blown features. I don't want to nitpick on naming but using different names might help here.

Otherwise, I fully agree with Stephen that we can either work on 1 feature at a time or we treat all features as one experiment.

Reply:
My concern with having a single preview space for all features is that we won't be able to track experimental interdependencies among our own bits.

Let's say the ASP.NET team decides to ship some preview APIs for Blazor. Say the work spans multiple assemblies. The top level assembly now has to be marked as "I need preview features", which is a good thing. But it also means that the implementer of this API might accidentally end up depending on some unrelated preview BCL API or language feature. Now let's say the team promised to ship a stable version in the next release, but the BCL/language feature isn't ready for that yet. Since we know communication across larger teams is hard, I'd say it's pretty plausible that we might find this out late, putting the productization of the Blazor APIs at risk.

The same is also true for customers. Just because I want to use some experimental EF features doesn't necessarily mean I'm OK with accidentally depending on some new type system feature which I now inadvertently base my entire architecture on.

I think it's totally fine to say:

1. The compiler cannot support multiple experimental features -- it only has a preview on/off mode.
2. The runtime may or may not support multiple preview features.
3. Library layers can have multiple independent preview features (the transitivity enforcement helps to detect overlap)

We have two options to address this:

1. We give up on modeling experimental compiler features via the `ExperimentalFeature` property and ask projects to set `<LangVersion>Preview</LangVersion>` when the compiler complaints.
2. We add some custom targets to the SDK that automatically does it when specific experimental features are turned on.

But I think giving up on modeling independent experimental features for the entire stack just because certain parts of the stack cannot sensibly support independent experiments feels like throwing out the baby with the bathwater and undermine a basic philosophy of this approach: ensuring customers don't accidentally on experimental features.

Reply:
> The compiler cannot support multiple experimental features -- it only has a preview on/off mode.

I think we'll find that the compiler is not special here. I think the problems we face will likely be replicated in other tools. Consider the SDK for a moment as it's reasonable that future experiments will drift into needing SDK changes. Given their complexity lI think it's plausible they end up in the same boat here: can't afford the test matrix of a series of on / off but can't avoid it because the feature impacts VS. 

> We give up on modeling experimental compiler features via the ExperimentalFeature property and ask projects to set <LangVersion>Preview</LangVersion> when the compiler complaints.

My expectation is that if we have some tools which are on / off and some which are granular then the on / off tools flip to "preview on" whenever any of the granular features are "on". Anything else is just pushing complexity to the user for very little win.

Reply:
> But I think giving up on modeling independent experimental features for the entire stack just because certain parts of the stack cannot sensibly support independent experiments feels like throwing out the baby with the bathwater

We already have ways of shipping experiments for most runtime work, e.g. environment variables.  We already have ways of shipping experiments for most library work, e.g. as an OOB package.  I expect the majority of the things where we actually want/need to ship them as part of the actual release in a visible way (as opposed to, say, runtimelab) are where it's part of a full stack change / improvement.  Shipping numerical interfaces is useless without the language support for statics in interfaces.  Shipping UTF8 support is significantly diminished without language support for UTF8 literals (multiple folks have told us it's the most important aspect of the feature to them).  Etc.  To me, this whole thing is about the entire stack.

Reply:
Roslyn has established system of feature branches, e.g. (sharplab)[https://sharplab.io/] allows you to play with them. It is not that hard to get Roslyn from feature branch combined with runtime from runtimelab, but it always going to be more hassle than just enabling a switch on top of .NET runtime that you have installed by default.

To me, this is primarily about barrier to entry to try the preview and about our own engineering efficiency (working in a branch for extended period of time is less efficient).